File: src/main/java/org/cascadebot/cascadebot/data/objects/donation/TimeFlag.java
package org.cascadebot.cascadebot.data.objects.donation;

import com.google.gson.JsonObject;
import lombok.Getter;
import org.cascadebot.cascadebot.data.language.Language;
import org.cascadebot.cascadebot.data.language.Locale;
import org.cascadebot.cascadebot.utils.FormatUtils;

public class TimeFlag extends DataFlag {

    private TimeFlag() {
        super();
    }

    public TimeFlag(String id) {
        super(id);
    }

    @Getter
    private long mills;

    @Override
    DataFlag parseFlagData(JsonObject flagDataObject) {
        mills = flagDataObject.get("time").getAsLong();
        return this;
    }

    @Override
    public String toString() {
        return "TimeFlag(id=" + getId() + ", time=" + mills + ")";
    }

    @Override
    public String getDescription(Locale locale) {
        return Language.i18n(locale, "flags." + id + ".description", FormatUtils.formatLongTimeMills(mills));
    }

}

File: src/main/kotlin/org/cascadebot/cascadebot/data/objects/donation/TimeFlag.kt
package org.cascadebot.cascadebot.data.objects.donation

import com.google.gson.JsonObject
import lombok.Getter
import org.cascadebot.cascadebot.data.language.Language.i18n
import org.cascadebot.cascadebot.data.language.Locale
import org.cascadebot.cascadebot.utils.FormatUtils

class TimeFlag : DataFlag {
    private var mills: Long = 0

    private constructor() : super() {}
    constructor(id: String?, scope: FlagScope?) : super(id, scope) {}

    override fun parseFlagData(flagDataObject: JsonObject): DataFlag {
        mills = flagDataObject["time"].asLong
        return this
    }

    override fun toString(): String {
        return "TimeFlag(id=$id, time=$mills)"
    }

    override fun getDescription(locale: Locale?): String? {
        return i18n(locale!!, "flags.$id.description", FormatUtils.formatLongTimeMills(mills))
    }

    override fun compareTo(flag: DataFlag): Int {
        if (flag !is TimeFlag) {
            throw UnsupportedOperationException("Cannot compare different types of flags")
        }
        return mills.compareTo(flag.mills);
    }
}